File: CS50/C/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
    // Promting the user for a card number
    long card = get_long("Number: ");

    // Declaring card1 so we can't lose the cards value while calculating it's lenght
    long card1 = card;
    int sum1 = 0;
    int sum2 = 0;

    // Measure the card's lenght
    int lenght = 0;
    while (card > 0)
    {
        long num = card % 10;
        card -= num;
        card = card / 10;
        lenght++;

        // Doing the checknum and storing the two sums in two variables
        if (lenght % 2 == 0)
        {
            int n = num * 2;
            sum1 += n / 10;
            sum1 += n % 10;
        }
        else
        {
            sum2 += num;
        }
    }
    int sum = sum1 + sum2;

    // Calculating the first 2 digits of the card
    long start = card1 / pow(10, (lenght - 2));
    long start1dig = start / 10;

    // Identifying which card is it and adding another statement to see if it's valid or not
    if (lenght == 15 && (start == 34 || start == 37) && (sum % 10 == 0))
    {
        printf("AMEX\n");
    }
    else if (lenght == 16 && (start == 51 || start == 52 || start == 53 || start == 54 || start == 55) && (sum % 10 == 0))
    {
        printf("MASTERCARD\n");
    }
    else if ((lenght == 13 || lenght == 16) && (start1dig == 4) && (sum % 10 == 0))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }

}

File: CS50/Python/credit/credit.py
"""
 credit.py
 Program that determinates if a credit card is valid and prints its' type
 By Pol
 Created 22/04/2022
"""
import sys


def main():
    # Promting the user for a card
    try:
        card = input("Number: ")
    except:
        sys.exit(1)

    # Determinating the cards' validity
    if card_valid(card) == False:
        print("INVALID")
        sys.exit(1)

    # Determinating the cards' type
    card_type(card)


# ____________________o_________________________
# Function that determinates if the card is valid or not with Luhn's Algorithm
def card_valid(card):
    lenght = len(card)
    card = int(card)
    count = 0
    sum = 0
    current_number = 0

    # Reading each number from the last number
    while card != 0:
        count += 1
        current_number = card % 10
        card = card // 10
        # If we're reading a second-to-last digit, we multiply by 2 and add
        if count % 2 == 0:
            current_number *= 2
            sum += current_number % 10
            sum += current_number // 10
        # Else, we only add
        else:
            sum += current_number

    # If the sum ends in 0, it is a valid card
    if sum % 10 == 0:
        return True
    else:
        return False

# ____________________o_________________________
# Function that determinates the cards type
def card_type(card):
    lenght = len(card)

    # Getting the credit cards' first two digits
    first_digits = int(int(card) // 10**(lenght - 2))

    # AMEX: 15 digits, starts with 34 or 37
    if (lenght == 15) and (first_digits in (34, 37)):
        print("AMEX")

    # MASTERCARD: 16 digits, starts with 51, 52, 53, 54 or 55
    if (lenght == 16) and (first_digits in (51, 52, 53, 54, 55)):
        print("MASTERCARD")

    # VISA: 13 or 16 digits, starts with 4
    if (lenght in (13, 16)) and (first_digits // 10 == 4):
        print("VISA")


# Calling main()
main()
